pkg/controller/snapshot_controller.go

```diff
@@ -0,0 +1,773 @@
+		if _, err = ctrl.clientset.VolumesnapshotV1alpha1().VolumeSnapshotContents().Create(snapshotContent); err == nil || apierrs.IsAlreadyExists(err) {
+			// Save succeeded.
+			if err != nil {
+				glog.V(3).Infof("volume snapshot data %q for snapshot %q already exists, reusing", snapshotContent.Name, snapshotKey(snapshot))
+				err = nil
+			} else {
+				glog.V(3).Infof("volume snapshot data %q for snapshot %q saved", snapshotContent.Name, snapshotKey(snapshot))
+			}
+			break
+		}
+		// Save failed, try again after a while.
+		glog.V(3).Infof("failed to save volume snapshot data %q for snapshot %q: %v", snapshotContent.Name, snapshotKey(snapshot), err)
+		time.Sleep(ctrl.createSnapshotContentInterval)
+	}
+
+	if err != nil {
+		// Save failed. Now we have a storage asset outside of Kubernetes,
+		// but we don't have appropriate volumesnapshotdata object for it.
+		// Emit some event here and try to delete the storage asset several
+		// times.
+		strerr := fmt.Sprintf("Error creating volume snapshot data object for snapshot %s: %v.", snapshotKey(snapshot), err)
+		glog.Error(strerr)
+		ctrl.eventRecorder.Event(newSnapshot, v1.EventTypeWarning, "CreateSnapshotContentFailed", strerr)
+		return nil, err
+	}
+
+	// save succeeded, bind and update status for snapshot.
+	result, err := ctrl.bindandUpdateVolumeSnapshot(snapshotContent, newSnapshot)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// Delete a snapshot
+// 1. Find the SnapshotContent corresponding to Snapshot
+//   1a: Not found => finish (it's been deleted already)
+// 2. Ask the backend to remove the snapshot device
+// 3. Delete the SnapshotContent object
+// 4. Remove the Snapshot from vsStore
+// 5. Finish
+func (ctrl *csiSnapshotController) DeleteSnapshotContentOperation(content *crdv1.VolumeSnapshotContent) error {
+	glog.V(4).Infof("deleteSnapshotOperation [%s] started", content.Name)
+
+	// get secrets if VolumeSnapshotClass specifies it
+	var snapshotterCredentials map[string]string
+	snapshotClassName := content.Spec.VolumeSnapshotClassName
+	if len(snapshotClassName) != 0 {
+		if snapshotClass, err := ctrl.clientset.VolumesnapshotV1alpha1().VolumeSnapshotClasses().Get(snapshotClassName, metav1.GetOptions{}); err == nil {
+			// Resolve snapshotting secret credentials.
+			// No VolumeSnapshot is provided when resolving delete secret names, since the VolumeSnapshot may or may not exist at delete time.
+			snapshotterSecretRef, err := GetSecretReference(snapshotterSecretNameKey, snapshotterSecretNamespaceKey, snapshotClass.Parameters, content.Name, nil)
+			if err != nil {
+				return err
+			}
+			snapshotterCredentials, err = GetCredentials(ctrl.client, snapshotterSecretRef)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	err := ctrl.handler.DeleteSnapshot(content, snapshotterCredentials)
+	if err != nil {
+		return fmt.Errorf("failed to delete snapshot %#v, err: %v", content.Name, err)
```

Comment:
Should we send event to the SnapshotContent so user (admin) can see what's wrong? This applies to all errors in this function, perhaps the error handling could be moved to `deleteSnapshotContent`.

Reply:
Added event.